### PR TITLE
APPS 3232 - Update status callback endpoint and trackUndeliveredMessage

### PR DESCRIPTION
**Pull request by bergieboy**

### CODE REVIEW

**Review by solomontang — DISMISSED** [review]

File: src/main/java/com/openlattice/chronicle/controllers/v2/ChronicleControllerV2.java [hunk 1]
```diff
@@ -180,6 +183,22 @@ public void sendMessages(
...
     }
 
+    @RequestMapping(
+            path = { ORGANIZATION_ID_PATH + MESSAGE_PATH + STATUS_PATH },
+            method = RequestMethod.POST )
+    @ResponseStatus( HttpStatus.OK )
+    public void updateMessageStatus(
+            @PathVariable ( ORGANIZATION_ID ) UUID organizationId,
+            HttpServletRequest request
+    ) throws ExecutionException {
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/controllers/v2/ChronicleControllerV2.java:193]: Following up on the feedback on https://github.com/openlattice/chronicle-api/pull/60

```suggestion
    public void updateMessageStatus(
            @PathVariable( ORGANIZATION_ID ) UUID organizationId,
            @RequestParam( MESSAGE_SID ) String messageSid,
            @RequestParam( MESSAGE_STATUS ) String messageStatus
    ) {
```

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/main/java/com/openlattice/chronicle/controllers/v2/ChronicleControllerV2.java [hunk 2]
```diff
@@ -180,6 +183,22 @@ public void sendMessages(
...
+        String messageSid = request.getParameter( "MessageId ");
+        String status = request.getParameter( "MessageStatus" );
+
+        if ( status.equals( "undelivered" ) || status.equals( "failed" ) ) {
+            messageService.trackUndeliveredMessage( organizationId, messageSid );
+        }
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/controllers/v2/ChronicleControllerV2.java:199]: `undelivered` and `failed` should be enums, perhaps `MessageStatus`? Also, if you create an enum, then the above suggestion would be ...

```
@RequestParam( MESSAGE_STATUS ) MessageStatus messageStatus	
```

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

**Review by UnsungHero97 — APPROVED** [review]

File: src/main/java/com/openlattice/chronicle/services/twilio/TwilioService.kt [hunk 3]
```diff
@@ -26,13 +29,21 @@ class TwilioService(configuration: TwilioConfiguration) :
...
                 .creator(PhoneNumber(messageDetails.phoneNumber), fromPhoneNumber, messageText)
-                .setStatusCallback(URI.create("https://api.openlattice.com/bifrost/messages/status"))
+                .setStatusCallback(URI.create(
+                    "https://api.openlattice.com$BASE/$organizationId$MESSAGE_PATH$STATUS_PATH"))
                 .create()
+            logger.info(
+                """
+                    message sent to participant ${messageDetails.participantId}
+                    for study ${messageDetails.studyId}
+                    in organization $messageDetails
```
Comment (anzioka) [inline src/main/java/com/openlattice/chronicle/services/twilio/TwilioService.kt:44]: should this be organizationId?

File: src/main/java/com/openlattice/chronicle/services/message/MessageService.kt [hunk 4]
```diff
@@ -30,11 +33,68 @@ open class MessageService(
...
+                }
+                try {
+                    Thread.sleep(1000) // if message sid not indexed in elasticsearch yet, wait a second and retry
+                } catch (e: InterruptedException) {
+                }
+                numAttempts++
+            }
+            if (entity != null) {
+                val entityKeyId = UUID.fromString(
+                        entity[EdmConstants.OL_EKID]!!.iterator().next().toString()
```
Comment (anzioka) [inline src/main/java/com/openlattice/chronicle/services/message/MessageService.kt:75]: no change required
`entity[EdmConstants.OL_EKID]!!` can be replaced with `entity.getValue(EdmConstants.OL_EKID) `

**Review by anzioka — DISMISSED** [review]

File: src/main/java/com/openlattice/chronicle/services/message/MessageService.kt [hunk 5]
```diff
@@ -30,11 +33,68 @@ open class MessageService(
...
+        protected val logger: Logger = LoggerFactory.getLogger(MessageService::class.java)
+        private const val MAX_FAILURE_ATTEMPTS = 5
+
+    }
+
+    open fun getSearchTermString(propertyTypeId: UUID, searchString: String): String {
+        return "entity.$propertyTypeId:\"$searchString\""
+    }
+
+    @Throws(ExecutionException::class)
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/message/MessageService.kt:45]: Why did you add this? You don't need it, plus the try-catch already catches it.

(You can also remove it from `recordMessagesSent`)

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/main/java/com/openlattice/chronicle/services/message/MessageService.kt (hunk 5 above)
Comment (bergieboy) [reply src/main/java/com/openlattice/chronicle/services/message/MessageService.kt:45]: I'm so confused, I didn't add it. but it was there locally.

Comment (bergieboy) [reply src/main/java/com/openlattice/chronicle/services/message/MessageService.kt:45]: I though it came from develop when I merged develop and resolved changes

Comment (UnsungHero97) [reply src/main/java/com/openlattice/chronicle/services/message/MessageService.kt]: There's another `@Throws` you can remove on line 100.

Comment (geekbeast) [reply src/main/java/com/openlattice/chronicle/services/message/MessageService.kt]: Kotlin conversion will often add these in.

**Review by anzioka — DISMISSED** [review]

File: src/main/java/com/openlattice/chronicle/controllers/v2/ChronicleControllerV2.java [hunk 6]
```diff
@@ -187,9 +179,18 @@ public void sendMessages(
         messageService.sendMessages( organizationId, messageDetails );
     }
 
-    @Override public void updateMessageStatus(
-            UUID organizationId, String messageId, MessageStatus messageStatus ) {
-        throw new NotImplementedException( "not implemented" );
+    @RequestMapping(
+            path = { ORGANIZATION_ID_PATH + MESSAGE_PATH + STATUS_PATH },
+            method = RequestMethod.POST )
+    @ResponseStatus( HttpStatus.NO_CONTENT )
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/controllers/v2/ChronicleControllerV2.java:185]: No change required. I don't see the benefit of `NO_CONTENT` over simply `OK`.

File: src/main/java/com/openlattice/chronicle/services/twilio/TwilioService.kt [hunk 7]
```diff
@@ -26,13 +29,21 @@ class TwilioService(configuration: TwilioConfiguration) :
...
+                .setStatusCallback(URI.create(
+                    "https://api.openlattice.com$BASE/$organizationId$MESSAGE_PATH$STATUS_PATH"))
                 .create()
+            logger.info(
+                """
+                    message sent to participant ${messageDetails.participantId}
+                    for study ${messageDetails.studyId}
+                    in organization $organizationId
+                """.trimIndent()
+            )
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/twilio/TwilioService.kt:46]: No change required. This will log a multi-line string which makes it harder to grep for, and possibly not what you intended.

**Review by UnsungHero97 — APPROVED** [review]

File: src/main/java/com/openlattice/chronicle/controllers/v2/ChronicleControllerV2.java (hunk 6 above)
Comment (bergieboy) [reply src/main/java/com/openlattice/chronicle/controllers/v2/ChronicleControllerV2.java:185]: this was in response to some errors that joe pointed out in our twillio logs a couple months ago. 

```
Twilio was unable to fetch content from: https://api.openlattice.com/bifrost/reminders/status
Error: Error reading response: Response does not contain content type
Account SID: AC51dec3a215872a12cfea7b6ce666953b
SID: SMe594701a52a54777a11c50d5ffc8a612
Request ID: 5bbae950-f591-4361-9074-ddd34770e42f
Remote Host: api.openlattice.com
Request Method: POST
Request URI: https://api.openlattice.com/bifrost/reminders/status
SSL Version: TLSv1.2
URL Fragment: true
```

**Review by UnsungHero97 — APPROVED** [review]

**Review by geekbeast — APPROVED** [review]
